content/blog/2019-10-25-efficient-path-profiling/index.md

```diff
@@ -0,0 +1,109 @@
+This path profiling algorithm computes information about paths directly, and so does not suffer from the same loss of accuracy as approximations derived from edge use measurements.
+As promised by the title, this algorithm is both memory efficient and fast enough to use when profiling realistic programs.
+Experimentally, path profiling is shown to only have a 31% profiling overhead compared with edge profiling having a 16% overhead on a standard benchmark suite.
+
+## The Algorithm for DAGs
+
+We start by assuming profiling over a directed acyclic graph (DAG), but will later explore extending to loops.
+Any DAG will have at least node with no incoming edges, called entry nodes, and at least node with no outgoing edges, called exit nodes.
+We extend any CFG with a single entry and single exit node, thus making each unique.
+The steps of the path profiling algorithm are as follows:
+1. Assign a minimal number of integer values to edges such that any calculated edge sum is unique.
+2. Using a spanning tree, create and select instrumentation for computing the increment for each edge.
+3. Collect the dynamic runtime profile
+4. Derive the executed paths based on the results and selected instrumentation.
+
+### Assigning Edge Integer Values
+
+The first step to path profiling is to assign edge integers such that any taken path in the DAG results in a unique integer value.
+There are many such assignments, but the minimal positive set can be given by reasoning about the number of paths each edge can take to reach the exit node.
+Intuitively, calculating a unique integer based on paths works since each split creates a new path from the unique entry node and results in a distinct integer.
+Formally, the value for each edge from _v_ to _w_ is constructed to be the sum of the number of paths from all previously examined nodes extending from _v_.
+This result, for example, in the integer values for our sample CFG:
+
+<img src="integers.png" alt="drawing" width="300"/>
+
+### Creating Instrumentation
+
+The unique paths given by these integers is not yet efficient counting instrumentation.
+Counting instrumentation is based on incrementing a specific register whenever a given edge is passed; such writes can be expensive, so we select edges to minimize the number of writes while still producing correct results.
+To construct counting efficiently, we must first identify the most taken edges (based on the edge weights computed earlier) and avoid incrementing when using those edges.
+Specifically, we compute the maximum spanning tree with respect to edge weights.
+All updates then only reason about those edges in the _chord_, the set of edges not in the maximum spanning tree.
+
+Instrumentation for the graph (choosing which register each path updates) is then added for each edge in the chord based on the integers computed earlier.
```

Comment:
More terminology silliness: the _instrumentation_ would not be the _choice_, it would be the extra code inserted to actually do the update. (This parenthetical sounds like the instrumentation is just about choosing what to do, not doing it.)